### ISSUE TITLE
Can't run migrations if using other modules in entities

### ISSUE THREAD

**Issue by robertnisipeanu** [mikro-orm/mikro-orm#383]
**Describe the bug**
I have an entity with some methods that uses an external module, because of that I can't run the migrations (I have to comment the code where I use the module to be able to, which is ok on local/dev machine, but on the prod machine where it is configured to auto-deploy from the git master branch, I can't do that).

**Stack trace**
```
{ Error: Cannot find module 'alt-server'
    at Function.Module._resolveFilename (internal/modules/cjs/loader.js:636:15)
    at Function.Module._load (internal/modules/cjs/loader.js:562:25)
    at Module.require (internal/modules/cjs/loader.js:692:17)
    at require (internal/modules/cjs/helpers.js:25:18)
    at Object.<anonymous> (/home/altv/actions-runner/_work/ggez_altv/ggez_altv/server/src/database/PlayerDataManager.ts:1:1)
    at Module._compile (internal/modules/cjs/loader.js:778:30)
    at Module.m._compile (/home/altv/actions-runner/_work/ggez_altv/ggez_altv/server/node_modules/ts-node/src/index.ts:814:23)
    at Module._extensions..js (internal/modules/cjs/loader.js:789:10)
    at Object.require.extensions.(anonymous function) [as .ts] (/home/altv/actions-runner/_work/ggez_altv/ggez_altv/server/node_modules/ts-node/src/index.ts:817:12)
    at Module.load (internal/modules/cjs/loader.js:653:32)
    at tryModuleLoad (internal/modules/cjs/loader.js:593:12)
    at Function.Module._load (internal/modules/cjs/loader.js:585:3)
    at Module.require (internal/modules/cjs/loader.js:692:17)
    at require (internal/modules/cjs/helpers.js:25:18)
    at Object.<anonymous> (/home/altv/actions-runner/_work/ggez_altv/ggez_altv/server/src/database/index.ts:2:1)
    at Module._compile (internal/modules/cjs/loader.js:778:30) code: 'MODULE_NOT_FOUND' }
```

**To Reproduce**
1. Add some methods that are not annotated with `@Property` or anything DB related
2. Use an external module inside that method
3. Try to run `npx mikro-orm migration:create` or `npx mikro-orm migration:up`

**Expected behavior**
That module should be ignored as there isn't anything related to a db(`@Property`) accessing the module and migrations should be run

**Additional context**
This is the mikro-orm config I'm using:
```
import {Options, ReflectMetadataProvider} from "mikro-orm";
import Account from "./_entities/Account";
import PlayerLevel from "./_entities/PlayerLevel";
import PlayerMoney from "./_entities/PlayerMoney";
import {MYSQL_DATABASE, MYSQL_HOST, MYSQL_PASSWORD, MYSQL_USERNAME} from "./utils/config";

const config = {
    metadataProvider: ReflectMetadataProvider,
    discovery: {disableDynamicFileAccess: true},
    cache: {enabled: false},
    entities: [Account, PlayerLevel, PlayerMoney],
    dbName: MYSQL_DATABASE,
    host: MYSQL_HOST,
    port: 3306,
    user: MYSQL_USERNAME,
    password: MYSQL_PASSWORD,
    type: "mysql"
} as Options;

export default config;
```
And this is my entity:
```
import Account from "./Account";
import {DatabaseEvents, DatabaseEventType} from "../database";
import {Entity, IdEntity, OneToOne, PrimaryKey, Property} from "mikro-orm";

@Entity()
export default class PlayerLevel implements IdEntity<PlayerLevel> {

    @PrimaryKey()
    id: number;

    @OneToOne(() => Account)
    user: Account;

    @Property({default: 1, name: 'level', type: 'number'})
    private _level: number = 1;

    get level(): number {
        return this._level;
    }

    set level(value: number) {
        if (this.user.playerObj)
            DatabaseEvents.getInstance().emit(DatabaseEventType.PlayerLevel.Level, this.user.playerObj, this._level, value);
        this._level = value;
    }


    @Property({default: 0, name: 'role_points', type: 'number'})
    private _rolePoints: number = 0;

    get rolePoints(): number {
        return this._rolePoints;
    }

    set rolePoints(value: number) {
        if (this.user.playerObj)
            DatabaseEvents.getInstance().emit(DatabaseEventType.PlayerLevel.RolePoints, this.user.playerObj, this._rolePoints, value);
        this._rolePoints = value;
    }

    @Property({name: 'created_at', onCreate: () => new Date(), type: 'date'})
    createdAt: Date;

    @Property({name: 'updated_at', onUpdate: () => new Date(), type: 'date'})
    updatedAt: Date;
}
```

**Versions**

| Dependency | Version |
| - | - |
| node | 10.16.3 |
| typescript | 3.7.5 |
| mikro-orm | 3.3.5 |
| mysql2 | 2.1.0 |

**Comment by B4nan** [comment]
Is `alt-server` a `devDependency`? 

> That module should be ignored as there isn't anything related to a db(@Property) accessing the module and migrations should be run

I dont understand why you say some module should be ignored. If it is inside the code, it won't be ignored. Running migrations cannot just ignore something. 

Also, where do you use that `alt-server`? It causes the error, yet you did not add any snippet where it is used. 

Anyway, I am pretty sure this is not related to MikroORM.

**Comment by B4nan** [comment]
I see `discovery: {disableDynamicFileAccess: true}`, so you are running via webpack? If so, you need to mark all external modules in your webpack config via `externals`. 

Something like this could work:

```
  externals: {
    knex: 'commonjs knex',
    umzug: 'commonjs umzug',
    'mikro-orm': 'commonjs mikro-orm',
    'alt-server': 'commonjs alt-server',
  },
```

With this you do not need that trick with `IgnorePlugin` and `optionalModules` described in the docs (will need to fix this, I discovered this just recently).

**Comment by robertnisipeanu** [comment]
I am not using webpack, however I've seen one issue with a similar problem and it was part of their config, so I tried to use it too.

What is the trick with `IgnorePlugin` and `optionalModules` as I do not find anything about that using the search function from the docs?

**Comment by B4nan** [comment]
Ignore that, its for webpack (and its here: https://mikro-orm.io/docs/deployment/#deploy-a-bundle-of-entities-and-dependencies-with-webpack). 

I am still waiting for the code snippet that is causing it as well as for the information about the module being a dev dependency.

**Comment by B4nan** [comment]
Btw what you claim just cannot be true. There are external dependencies used in entities in ORM tests, just like `uuid` or `mongodb`. 

Also, doing a search for that package on NPM does not return anything. So it is not a real dependency, right? 

https://www.npmjs.com/search?q=alt-server

**Comment by robertnisipeanu** [comment]
It is "injected" by the application that runs this one (present at runtime), so while it is not installed as a dependency, I have installed the typings as a dev dependency, which can be found here: https://www.npmjs.com/package/@altv/types

Even tho it is not able to find that while using `npx mikro-orm migration:[...]`, as it is not injected if running it like that, I think the migration maker just ignore imports that are only used in methods/properties not present with a `@Property`/`@PrimaryKey`/[...] decorators.

**Comment by B4nan** [comment]
> It is "injected" by the application that runs this one (present at runtime), so while it is not installed as a dependency, I have installed the typings as a dev dependency, which can be found here: https://www.npmjs.com/package/@altv/types

Typings are not enough, the error comes from actual import of that "library". What does it mean "injected", how do you do that? You basically need to do the same when running migrations (which you could probably do by adding that "injection" code to the ORM config as that will be required when running the CLI command). 

> I think the migration maker should just ignore imports that are only used in methods/properties not present with a @Property/@PrimaryKey/[...] decorators.

Well afaik this is just not possible, `node` (or `ts-node`) is running the code, and this error comes from it. I do not see a way how we could ignore deps. Not talking that I would see that as unwanted error hiding. Imagine that dependency would be actually needed to run the migration. I can't just ignore what was not found.

**Comment by robertnisipeanu** [comment]
> Typings are not enough, the error comes from actual import of that "library". What does it mean "injected", how do you do that? You basically need to do the same when running migrations (which you could probably do by adding that "injection" code to the ORM config as that will be required when running the CLI command).

There is a main application which I do not have access to it's source code, so I don't really know how it injects it. It calls mine like a submodule of it, while injecting it's own API into it.

>  Imagine that dependency would be actually needed to run the migration. I can't just ignore what was not found.
This is why I said if it is possible to not ignore things used in properties/methods decorated with `@Property` or any decorator that actually takes part into the migration.

Well, I think the way to fix it is to run the migrations at runtime in the code (where the dependency is also present and already injected). I will try this and come back with an answer.

**Comment by B4nan** [comment]
> Well, I think the way to fix it is to run the migrations at runtime in the code (where the dependency is also present and already injected). I will try this and come back with an answer.

Agreed, that sounds like the only way if you do not have control over it. You can use the `Migrator` directly. 

https://mikro-orm.io/docs/migrations/#using-the-migrator-programmatically

**Comment by robertnisipeanu** [comment]
Trying to do the migrations programmatically, however now I've got an issue where the app doesn't pick a name for the migration (that may be because when it is running, the TS is already compiled to JS? and it doesn't reads the classname anymore?).
```
(node:352440) UnhandledPromiseRejectionWarning: Error: Table 'altv.undefined' doesn't exist
    at Packet.asError (C:\Others\Projs\ggez_altv\server\build\node_modules\mysql2\lib\packets\packet.js:712:17)
    at Query.execute (C:\Others\Projs\ggez_altv\server\build\node_modules\mysql2\lib\commands\command.js:28:26)
    at Connection.handlePacket (C:\Others\Projs\ggez_altv\server\build\node_modules\mysql2\lib\connection.js:417:32)
    at PacketParser.onPacket (C:\Others\Projs\ggez_altv\server\build\node_modules\mysql2\lib\connection.js:75:12)
    at PacketParser.executeStart (C:\Others\Projs\ggez_altv\server\build\node_modules\mysql2\lib\packet_parser.js:75:16)
    at Socket.<anonymous> (C:\Others\Projs\ggez_altv\server\build\node_modules\mysql2\lib\connection.js:82:25)
    at Socket.emit (events.js:200:13)
    at Socket.EventEmitter.emit (domain.js:471:20)
    at addChunk (_stream_readable.js:294:12)
    at readableAddChunk (_stream_readable.js:275:11)
    at Socket.Readable.push (_stream_readable.js:210:10)
    at TCP.onStreamRead (internal/stream_base_commons.js:166:17)
    at emitWarning (internal/process/promises.js:120:15)
    at processPromiseRejections (internal/process/promises.js:168:7)
    at processTicksAndRejections (internal/process/task_queues.js:90:32)
(node:352440) Error: Table 'altv.undefined' doesn't exist
    at Packet.asError (C:\Others\Projs\ggez_altv\server\build\node_modules\mysql2\lib\packets\packet.js:712:17)
    at Query.execute (C:\Others\Projs\ggez_altv\server\build\node_modules\mysql2\lib\commands\command.js:28:26)
    at Connection.handlePacket (C:\Others\Projs\ggez_altv\server\build\node_modules\mysql2\lib\connection.js:417:32)
    at PacketParser.onPacket (C:\Others\Projs\ggez_altv\server\build\node_modules\mysql2\lib\connection.js:75:12)
    at PacketParser.executeStart (C:\Others\Projs\ggez_altv\server\build\node_modules\mysql2\lib\packet_parser.js:75:16)
    at Socket.<anonymous> (C:\Others\Projs\ggez_altv\server\build\node_modules\mysql2\lib\connection.js:82:25)
    at Socket.emit (events.js:200:13)
    at Socket.EventEmitter.emit (domain.js:471:20)
    at addChunk (_stream_readable.js:294:12)
    at readableAddChunk (_stream_readable.js:275:11)
    at Socket.Readable.push (_stream_readable.js:210:10)
    at TCP.onStreamRead (internal/stream_base_commons.js:166:17)
(node:352440) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
    at emitDeprecationWarning (internal/process/promises.js:134:13)
    at emitWarning (internal/process/promises.js:127:3)
    at processPromiseRejections (internal/process/promises.js:168:7)
    at processTicksAndRejections (internal/process/task_queues.js:90:32)
```

**Comment by B4nan** [comment]
Name for migration? The error is about wrong table name, so I guess you mean entity name? But in your example you are not specifying the name manually, so not sure where that prefix is coming from? 

If you are running compiled code via `node`, they you should have the migration configuration adjusted that way too (`path` should lead to `dist` and `pattern` needs to be adjusted to execute JS, as by default it will look for TS files. 

Without a reproduction repo there is not much I can do for you, especially in such a nonstandard environment...

**Comment by robertnisipeanu** [comment]
The `migration.path` already points to the /dist path, however the pattern is not modified. What does the pattern needs to be exactly? It needs to pick up all .js entities?

**Comment by B4nan** [comment]
No, it's not about entities, its about migration files - those that sits in the `migrations.path`. This is the default: `pattern: /^[\w-]+\d+\.ts$/,`, so as long as you will keep the generated name, just change the suffix: `pattern: /^[\w-]+\d+\.js$/,`.

But your error is about database table, and I do not have a way to know which table. Do you configure the `altv.` prefix somewhere?

**Comment by B4nan** [comment]
You could try to enable async stack traces to see where this is coming from: http://knexjs.org/#Installation-asyncStackTraces

    driverOptions: { asyncStackTraces: true },

**Comment by robertnisipeanu** [comment]
No, `altv` is the database name, it's not a prefix.
That error is triggered when I run this:
```
 MikroORM.init(config).then(async (orm) => {

     const migrator = orm.getMigrator();
     const test = await migrator.createMigration();
     alt.log("Created migration: " + test.fileName);
     await migrator.up();

     new Main(orm);
 });
```
It triggers at this line:
` const test = await migrator.createMigration();`

I do have the project in a private repo, if you do want, I can give you temporary access to see how I have my MikroORM entities/config setup.

**Comment by robertnisipeanu** [comment]
Got the `migrator.createMigration()` think to work on a local dev machine. However, now I'm facing this issue:
```
[20:23:19] Created migration: Migration20200305182319.ts
== Migration20200305182158: migrating =======
(node:15664) UnhandledPromiseRejectionWarning: D:\GGEZ\ggez_altv\server\build\src\migrations\Migration20200305182158.ts:1
import { Migration } from 'mikro-orm';
       ^

SyntaxError: Unexpected token {
    at Module._compile (internal/modules/cjs/loader.js:722:23)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:798:10)
    at Module.load (internal/modules/cjs/loader.js:645:32)
    at Function.Module._load (internal/modules/cjs/loader.js:560:12)
    at Module.require (internal/modules/cjs/loader.js:685:19)
    at require (internal/modules/cjs/helpers.js:16:16)
    at Migrator.resolve (D:\GGEZ\ggez_altv\server\build\node_modules\mikro-orm\dist\migrations\Migrator.js:65:27)
    at Object.customResolver (D:\GGEZ\ggez_altv\server\build\node_modules\mikro-orm\dist\migrations\Migrator.js:26:46)
    at Migration.migration (D:\GGEZ\ggez_altv\server\build\node_modules\umzug\lib\migration.js:63:38)
    at D:\GGEZ\ggez_altv\server\build\node_modules\umzug\lib\migration.js:123:37
    at Generator.next (<anonymous>)
    at step (D:\GGEZ\ggez_altv\server\build\node_modules\babel-runtime\helpers\asyncToGenerator.js:17:30)
    at D:\GGEZ\ggez_altv\server\build\node_modules\babel-runtime\helpers\asyncToGenerator.js:35:14
    at new Promise (<anonymous>)
    at new F (D:\GGEZ\ggez_altv\server\build\node_modules\core-js\library\modules\_export.js:36:28)
    at D:\GGEZ\ggez_altv\server\build\node_modules\babel-runtime\helpers\asyncToGenerator.js:14:12
    at emitWarning (internal/process/promises.js:120:15)
    at processPromiseRejections (internal/process/promises.js:168:7)
    at processTicksAndRejections (internal/process/task_queues.js:90:32)
(node:15664) D:\GGEZ\ggez_altv\server\build\src\migrations\Migration20200305182158.ts:1
import { Migration } from 'mikro-orm';
       ^

SyntaxError: Unexpected token {
    at Module._compile (internal/modules/cjs/loader.js:722:23)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:798:10)
    at Module.load (internal/modules/cjs/loader.js:645:32)
    at Function.Module._load (internal/modules/cjs/loader.js:560:12)
    at Module.require (internal/modules/cjs/loader.js:685:19)
    at require (internal/modules/cjs/helpers.js:16:16)
    at Migrator.resolve (D:\GGEZ\ggez_altv\server\build\node_modules\mikro-orm\dist\migrations\Migrator.js:65:27)
    at Object.customResolver (D:\GGEZ\ggez_altv\server\build\node_modules\mikro-orm\dist\migrations\Migrator.js:26:46)
    at Migration.migration (D:\GGEZ\ggez_altv\server\build\node_modules\umzug\lib\migration.js:63:38)
    at D:\GGEZ\ggez_altv\server\build\node_modules\umzug\lib\migration.js:123:37
    at Generator.next (<anonymous>)
    at step (D:\GGEZ\ggez_altv\server\build\node_modules\babel-runtime\helpers\asyncToGenerator.js:17:30)
    at D:\GGEZ\ggez_altv\server\build\node_modules\babel-runtime\helpers\asyncToGenerator.js:35:14
    at new Promise (<anonymous>)
    at new F (D:\GGEZ\ggez_altv\server\build\node_modules\core-js\library\modules\_export.js:36:28)
    at D:\GGEZ\ggez_altv\server\build\node_modules\babel-runtime\helpers\asyncToGenerator.js:14:12
(node:15664) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
    at emitDeprecationWarning (internal/process/promises.js:134:13)
    at emitWarning (internal/process/promises.js:127:3)
    at processPromiseRejections (internal/process/promises.js:168:7)
    at processTicksAndRejections (internal/process/task_queues.js:90:32)


```

Also, is there a way to set it to generate a .js migration (as it'll run when it's builded in js format and it's not ts anymore).

Here is the migration it generates:
```
import { Migration } from 'mikro-orm';

export class Migration20200305182319 extends Migration {

  async up(): Promise<void> {
    this.addSql('create table `account` (`id` int unsigned not null auto_increment primary key, `username` varchar(255) not null, `password` varchar(255) not null, `created_at` datetime not null, `updated_at` datetime not null) default character set utf8 engine = InnoDB;');

    this.addSql('create table `player_level` (`id` int unsigned not null auto_increment primary key, `user_id` int(11) unsigned not null, `level` int(11) not null default 1, `role_points` int(11) not null default 0, `created_at` datetime not null, `updated_at` datetime not null) default character set utf8 engine = InnoDB;');
    this.addSql('alter table `player_level` add index `player_level_user_id_index`(`user_id`);');
    this.addSql('alter table `player_level` add unique `player_level_user_id_unique`(`user_id`);');

    this.addSql('alter table `player_level` add constraint `player_level_user_id_foreign` foreign key (`user_id`) references `account` (`id`) on update cascade;');

    this.addSql('create table `player_money` (`id` int unsigned not null auto_increment primary key, `user_id` int(11) unsigned not null, `money` int(11) not null default 0, `bank_money` int(11) not null default 0, `created_at` datetime not null, `updated_at` datetime not null) default character set utf8 engine = InnoDB;');
    this.addSql('alter table `player_money` add index `player_money_user_id_index`(`user_id`);');
    this.addSql('alter table `player_money` add unique `player_money_user_id_unique`(`user_id`);');

    this.addSql('alter table `player_money` add constraint `player_money_user_id_foreign` foreign key (`user_id`) references `account` (`id`) on update cascade;');
  }

}
```

**Comment by B4nan** [comment]
That is because you are executing via `node` but the migration it finds is a TS file. Did you change the `path` and `pattern` so reflect you want to execute the compiled version?

> Also, is there a way to set it to generate a .js migration (as it'll run when it's builded in js format and it's not ts anymore).

Currently not, but going forward I would like to support JS migrations too. Will be happy to accept PR.

**Comment by B4nan** [comment]
You might be able to get around this by enabling ts-node. Should be enough to have it installed and enable `useTsNode` flag in package.json.

**Comment by robertnisipeanu** [comment]
I'm gonna try and modify the MigrationGenerator to generate the migrations depending on what is inside the option "migrations"
Is something like this ok for you?:
```
export type MigrationsOptions = {
  tableName?: string;
  path?: string;
  pattern?: RegExp;
  transactional?: boolean;
  disableForeignKeys?: boolean;
  allOrNothing?: boolean;
  javascriptMode?: boolean;
};
```
With the default value being false (for typescript generation)?
If javascriptMode is set to true then it would generate the migration file for CommonJS.
Also, while trying to build mikro-orm I'm getting quite some error, did I miss something?
```
D:\GGEZ\mikro-orm (master -> origin) (mikro-orm@3.3.5)
λ npm run build

> mikro-orm@3.3.5 build D:\GGEZ\mikro-orm
> rimraf dist && tsc

lib/connections/MongoConnection.ts:132:37 - error TS2344: Type 'T' does not satisfy the constraint '{ _id: any; }'.

132     let res: InsertOneWriteOpResult<T> | UpdateWriteOpResult | DeleteWriteOpResultObject | number;
                                        ~

lib/migrations/Migrator.ts:51:5 - error TS2322: Type 'Migration[]' is not assignable to type 'UmzugMigration[]'.
  Property 'path' is missing in type 'Migration' but required in type 'UmzugMigration'.

51     return this.umzug.pending();
       ~~~~~~~~~~~~~~~~~~~~~~~~~~~~

  lib/migrations/Migrator.ts:124:32
    124 export type UmzugMigration = { path: string; file: string };
                                       ~~~~
    'path' is declared here.

lib/migrations/Migrator.ts:55:5 - error TS2322: Type 'Migration[]' is not assignable to type 'UmzugMigration[]'.

55     return this.runMigrations('up', options);
       ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

lib/migrations/Migrator.ts:59:5 - error TS2322: Type 'Migration[]' is not assignable to type 'UmzugMigration[]'.

59     return this.runMigrations('down', options);
       ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~


Found 4 errors.

npm ERR! code ELIFECYCLE
npm ERR! errno 2
npm ERR! mikro-orm@3.3.5 build: `rimraf dist && tsc`
npm ERR! Exit status 2
npm ERR!
npm ERR! Failed at the mikro-orm@3.3.5 build script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     C:\Users\Robert\AppData\Roaming\npm-cache\_logs\2020-03-05T19_10_03_562Z-debug.log
```

**Comment by B4nan** [comment]
I would go for `emit: 'js' | 'ts';`, ideally we should base the default value on ts-node support being detected or forced via `tsNode` flag. But for simplicity let's start with `'ts'` as the default. 

That build error might be some dependency mismatch (maybe TS version? but that should be locked), try to use `yarn` so you install based on the lock file.

**Comment by robertnisipeanu** [comment]
Solved in #384 so I will close this issue.
